### PR TITLE
[10.0][sale_order_lot_mrp] add sale order lot mrp

### PR DESCRIPTION
**Pull request by mourad-ehm**

### CODE REVIEW

**Comment by mourad-ehm** [comment]
Hi @bealdav, @florian-dacosta,
Could you review this PR please.

**Comment by bealdav** [comment]
:+1:

**Comment by mourad-ehm** [comment]
Thanks @bealdav, 
ping @florian-dacosta

File: sale_order_lot_selection/model/stock.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ class StockMove(models.Model):
     _inherit = 'stock.move'
 
     @api.model
-    def _prepare_procurement_from_move(self, move):
-        vals = super(StockMove, self)._prepare_procurement_from_move(move)
-        vals['lot_id'] = move.restrict_lot_id.id
+    def _prepare_procurement_from_move(self):
+        vals = super(StockMove, self)._prepare_procurement_from_move()
+        vals['lot_id'] = self.restrict_lot_id.id
```
Comment (florian-dacosta) [inline sale_order_lot_selection/model/stock.py]: It should be api.multi else, this line can't work, right?

File: sale_order_lot_mrp/models/mrp.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            super(MrpProduction, production).button_plan()
+
+    def _generate_finished_moves(self):
+        move = super(MrpProduction, self).\
+            _generate_finished_moves()
+        move.restrict_lot_id = self.lot_id.id
+        return move
+
+
+class ProcurementOrder(models.Model):
```
Comment (florian-dacosta) [inline sale_order_lot_mrp/models/mrp.py]: It would be nice to have separate files for the different models. Not necessarily one model per file, but in any case, ProcurementOrder should not be in mrp.py file.

File: sale_order_lot_mrp/models/mrp.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                    if product.auto_generate_prodlot:
+                        idx += 1
+                        vals = production.lot_id.copy_data()[0]
+                        vals.update(self._get_custom_lot_vals(vals, idx))
+                        prodlot = prodlot_obj.create(vals)
+                        line_data['lot_id'] = prodlot
+        return boms_done, lines_done
+
+
+class StockProductionLot(models.Model):
```
Comment (florian-dacosta) [inline sale_order_lot_mrp/models/mrp.py]: Same here, I don't think this model belong in the mrp.py file.

**Review by florian-dacosta — COMMENTED** [review]
It seems good overall, I will approve once you check the decorator problem

File: sale_order_lot_selection/model/stock.py (hunk 1 above)
Comment (mourad-ehm) [reply sale_order_lot_selection/model/stock.py]: This method is defined without api.multi (https://github.com/odoo/odoo/blob/10.0/addons/stock/models/stock_move.py#L510). 
It is used in for loop (https://github.com/odoo/odoo/blob/10.0/addons/stock/models/stock_move.py#L488).

Comment (florian-dacosta) [reply sale_order_lot_selection/model/stock.py]: Yes, So I guess you should remove the api.model, to be coherent with native method

Comment (florian-dacosta) [reply sale_order_lot_selection/model/stock.py]: It may work this way, but it is confusing, usually we use `@api.model` when we don't care the value of `self`. But here, we need `self.restrict_lot_id.id`, so we actually care about the value.
That is why api.model is not good in this case.

Comment (mourad-ehm) [reply sale_order_lot_selection/model/stock.py]: OK. I removed api.model

**Review by florian-dacosta — APPROVED** [review]

**Comment by mourad-ehm** [comment]
thanks @florian-dacosta  for review.

**Comment by mourad-ehm** [comment]
Hi @sebastienbeau, could you merge this PR please ?

**Comment by bealdav** [comment]
@mourad-ehm please resolved conflicts

**Comment by rafaelbn** [comment]
Hi @mourad-ehm , please if you attend conflicts we can review and get this merge, thanks! @bealdav @florian-dacosta any additional comments?

**Comment by bealdav** [comment]
no additional comments excepts the v8 and v9 version of these modules are in production since several years.
Thanks for your review

File: sale_order_lot_mrp/README.rst [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+Allow you to attach a lot in manufacturing order
+that was defined by sale_order_lot_generator module
+for all concerned products (Lot Generation Auto field)
+Name of the MO is also based on sale name.
+
+Usage
+=====
+
+To use this module, you need to:
+
```
Comment (rafaelbn) [inline sale_order_lot_mrp/README.rst]: I would be very nice if you introduce how to use this module for testing. When you leave this section empty this increase the efforts of tested and in the most of cases they will leave the revision

**Comment by rafaelbn** [comment]
@bealdav I see:

- https://github.com/OCA/sale-workflow/tree/9.0/sale_order_lot_selection
- https://github.com/OCA/sale-workflow/tree/8.0/sale_order_lot_selection
- https://github.com/OCA/sale-workflow/tree/10.0/sale_order_lot_selection

So this is a NEW module: sale_order_lot_mrp. Isn't it?

**Comment by bealdav** [comment]
yes a new module in OCA, but lower versions are in production from akretion repo

File: sale_order_lot_mrp/README.rst (hunk 4 above)
Comment (antoniocanovas) [reply sale_order_lot_mrp/README.rst]: It does nothing on runbot ... I can't validate.

Comment (mourad-ehm) [reply sale_order_lot_mrp/README.rst]: I'll add more explanation on the use case off this module on readme file.

Comment (mourad-ehm) [reply sale_order_lot_mrp/README.rst]: I found a bug when validation a stock picking (after manufacturing products and generating lot number). I'll fix it.

**Comment by bealdav** [comment]
:+1:

File: sale_order_lot_mrp/README.rst (hunk 4 above)
Comment (mourad-ehm) [reply sale_order_lot_mrp/README.rst]: Hi @antoniocanovas, I fixed the bug and added use case on readme.

File: sale_order_lot_mrp/models/mrp.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                if mv_lot.product_id.tracking != 'none' and\
+                        mv_lot.product_id.auto_generate_prodlot and\
+                        workorder.final_lot_id:
+                    idx += 1
+                    vals = workorder.final_lot_id.copy_data()[0]
+                    vals.update(self._get_custom_lot_vals(vals, idx))
+                    vals.update({'product_id': mv_lot.product_id.id})
+                    prodlot = prodlot_obj.create(vals)
+                    mv_lot.lot_id = prodlot
+                    print prodlot.name
```
Comment (LoisRForgeFlow) [inline sale_order_lot_mrp/models/mrp.py]: remove `print` statements

File: sale_order_lot_mrp/models/stock.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            action_explode()
+        if not self.env.context.get('subcall', False) and processed_moves and \
+                original_lot:
+            index = 1
+            for new_move in processed_moves:
+                if new_move != self:
+                    vals = new_move._get_lot_vals(original_lot, index)
+                    lot = original_lot.copy(vals)
+                    new_move.restrict_lot_id = lot
+                    print lot.name
```
Comment (LoisRForgeFlow) [inline sale_order_lot_mrp/models/stock.py]: remove `print` statements

File: sale_order_lot_selection/model/stock.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 # @author Valentin CHEMIERE <valentin.chemiere@akretion.com>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import api, models
+from odoo import models, api
```
Comment (LoisRForgeFlow) [inline sale_order_lot_selection/model/stock.py]: Why this change, it was good alphabetically.

File: sale_order_lot_mrp/README.rst [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+Allow you to attach a lot in manufacturing order
+that was defined by sale_order_lot_generator module
+for all concerned products ("Lot Generation Auto" = True)
+Name of the MO is also based on sale name.
+
+Usage
+=====
+
+#. First define a bom kit that is composed by components wich are tracked
+by lot and have the option "Lot Generation Auto" = True.
```
Comment (LoisRForgeFlow) [inline sale_order_lot_mrp/README.rst]: Add 3 spaces to align with the above start of the sentences so this will show properly.

File: sale_order_lot_mrp/README.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+|
+
+.. image:: static/description/so_mrp_lot-4.png
+    :width: 800 px
+    :alt: Product Laptop Customized configuration
+
+|
+
+Create a sale order with for the KIT desk & laptop and confirm it.
```
Comment (LoisRForgeFlow) [inline sale_order_lot_mrp/README.rst]: Continue with the enumerated list `#.`

File: sale_order_lot_mrp/README.rst [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+|
+
+.. image:: static/description/so_mrp_lot-2.png
+    :width: 800 px
+    :alt: BOM KIT desk & laptop configuration
+
+|
+
+.. image:: quick_purchase/static/description/so_mrp_lot-3.png
```
Comment (LoisRForgeFlow) [inline sale_order_lot_mrp/README.rst]: not displayed

File: sale_order_lot_mrp/README.rst [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    :alt: Product Laptop Customized configuration
+
+|
+
+Create a sale order with for the KIT desk & laptop and confirm it.
+A lot number will be generated for this product.
+
+|
+
+.. image:: quick_purchase/static/description/so_mrp_lot-5.png
```
Comment (LoisRForgeFlow) [inline sale_order_lot_mrp/README.rst]: not displayed

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Code review

**Comment by LoisRForgeFlow** [comment]
Also I found the following issue in runbot:

image

When the reference was actually spare.

Not related to this module though, I would swear that another module introduces that feature, anyone remembers that module's name to check the issue?

**Comment by mourad-ehm** [comment]
This issue comes probably from the module "sale_order_revision"  not from "sale_order_lot_mrp".

**Comment by mourad-ehm** [comment]
HI @lreficent, I don't manage revision in my module this error comes from this module https://github.com/OCA/sale-workflow/blob/10.0/sale_order_revision/models/sale_order.py#L45. 
Can you check this please ?

**Comment by bealdav** [comment]
@lreficent could you approve now ? Thanks a lot

**Review by bealdav — APPROVED** [review]

**Review by LoisRForgeFlow — APPROVED** [review]
code review :+1:
